course3/homework-string/my_string.cc

```diff
@@ -0,0 +1,63 @@
+#include "my_string.h"
+#include <cstring>
+#include <stdexcept>
+
+MyString::MyString(const char* str) : buf(std::make_unique<char[]>(std::strlen(str)+1)) {
+	std::strcpy(buf.get(), str);
+}
+
+MyString::MyString(const MyString& obj) : buf(std::make_unique<char[]>(obj.size()+1)) {
+	std::strcpy(buf.get(), obj.buf.get());
+}
+
+bool MyString::operator==(const MyString& rhs) const {
+	if (!rhs.buf xor !buf)
+		return false;
+	if (!rhs.buf and !buf)
+		return true;
+	return std::strcmp(buf.get(), rhs.buf.get()) == 0;
+}
+
+bool MyString::operator!=(const MyString& rhs) const {
+	return !operator==(rhs);
+}
+
+const char& MyString::operator[](int i) const {
+	if (!buf)
+		throw std::range_error("Trying to index unallocated buffer.");
```

Comment:
This isn't entirely valid.
Either you decide that your ```[]``` does boundary checks, and then you check both the existence of buffer, and the bounds of the index (which would be easier if index is if unsigned type), or you don't do any checks at all.

Please note: the standard containers (```std::string```, ```std::vector```) don't perform any checks on ```operator[]```, but offer alternative function, ```at(size_t)``` that does the checks.